example/quantization/imagenet_gen_qsym_mkldnn.py

```diff
@@ -140,8 +140,8 @@ def save_params(fname, arg_params, aux_params, logger=None):
                              ' thresholds. This mode is expected to produce the best inference accuracy of all three'
                              ' kinds of quantized models if the calibration dataset is representative enough of the'
                              ' inference dataset.')
-    parser.add_argument('--quantized-dtype', type=str, default='uint8',
-                        choices=['int8', 'uint8'],
+    parser.add_argument('--quantized-dtype', type=str, default='auto',
```

Comment:
Can we somewhere document the benefits of using int8 versus uint8?

Reply:
Good idea. It's a trade off between int8 and uint8. One line suggestion is, we should use uint8 as much as we could to get best performance and accuracy. Int8 is better used for negitive inputs only. That's why we introduced `auto` out_type.

Reply:
Thanks for the explanation.  I'm still curious about why uint8 actually requires positive inputs.  Wouldn't an input array get the same affine transformation (scale and zero shift) applied to it just as the weights have before a fused kernel call?  I'll take a look at the FC gemm you're calling and see if I can better understand why we need int8 for negative inputs.  In any case not blocking for this PR.

Reply:
So reading a bit more about this from https://intel.github.io/mkl-dnn/ex_int8_simplenet.html 

```C++
auto user_src_memory = memory({ { { conv_src_tz }, memory::data_type::f32, memory::format::nchw }, cpu_engine }, user_src.data());
auto conv_src_md = memory::desc( { conv_src_tz }, memory::data_type::u8, memory::format::any);
auto conv_desc = convolution_forward::desc(prop_kind::forward,
        convolution_direct, conv_src_md, conv_weights_md, conv_bias_md,
        conv_dst_md, conv_strides, conv_padding, conv_padding,
        padding_kind::zero);
auto conv_prim_desc = convolution_forward::primitive_desc(conv_desc, conv_attr, cpu_engine);
auto conv_src_memory = memory(conv_prim_desc.src_primitive_desc());
auto src_reorder_pd = reorder::primitive_desc(user_src_memory.get_primitive_desc(), conv_src_memory.get_primitive_desc(), src_attr);
```

am I understanding correctly that we can convert from an fp32 input (with negative values) to uint8 via a reorder primitive.  Do we do this for the first negative input into subgraphs handled by mkldnn?

Reply:
@KellenSunderland Deep understanding of MKLDNNquantization is always welcome, which will help to use it better eventually. For your question, the answer is simple. MKLDNN transform data from fp32 to uint8 with scale and **zero shift**, which means, any negative value will be cut off to zero, it's equivalent to relu + quantize.

Reply:
Thanks a ton for the explanation @ZhennanQin.  I can see how from the users' point of view it makes sense to describe the dtype usage that way.  Am I correct in understanding that if you wanted to strictly use uint8 for resnets for example you could first quantize the input to uint8 and then not ignore the first convolution (because your inputs would already by positive?).  You'd have to understand your proper scale and zero-shift for input in this case.

Reply:
@KellenSunderland If data loader is configured with rgb_mean=(0,0,0), then the in_data for first convolution are all non-negative values, we can directly quantize the input to uint8 and don't need to ignore the first convolution. The flow will look like
input ->quantize->convolution(uint8)
If data loader is configured with rgb_mean!=(0,0,0), then in_data for first convolution will have negative values. For this case, you need to exclude the first convolution to run it in fp32 mode, and after the relu following first convolution, do quantization to uint8. The flow will look like,
input ->convolution(fp32) -> relu(fp32) ->quantize->pooling(uint8)